[DO NOT MERGE] Always prespecialize generic metadata.

nate-chandler opened the pull request:
Leaving open to benchmark and test source compatibility of turning on metadata prespecialization always, independent of target version.

nate-chandler committed:
[frontend] Default on metadata prespecialization.

Previously, -Xfrontend -prespecialize-generic-metadata had to be passed
in order for generic metadata to be prespecialized.  Now it is
prespecialized unless -Xfrontend
-disable-generic-metadata-prespecialization is passed.

nate-chandler commented:
@swift-ci please smoke benchmark

slavapestov commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -1132,6 +1133,13 @@ void IRGenerator::emitLazyDefinitions() {
       emitLazyTypeMetadata(*IGM.get(), type);
     }
+    while (!LazySpecializedTypeMetadataRecords.empty()) {
+      CanBoundGenericType type =
```
What about pre-specializing types such as `Foo<T>.Bar`?

nate-chandler replied:
Should work now.

slavapestov commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -3595,8 +3609,10 @@ llvm::GlobalValue *IRGenModule::defineTypeMetadata(CanType concreteType,
+  // Don't define the alias for foreign type metadata or prespecialized generic
+  // metadata, since neither is ABI.
+  if ((nominal && requiresForeignTypeMetadata(nominal)) ||
+      concreteType->is<BoundGenericType>())
```
Another suspicious use of BoundGenericType.

nate-chandler replied:
Yeah, that limitation is everywhere at the moment.  Would it be reasonable to start out only prespecializing these cases and expand from here?

nate-chandler commented:
Build failure:

```
18:29:59 FAILED: lib/swift/macosx/x86_64/libswiftRemoteMirror.dylib 
18:29:59 : && /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-perf/Ninja-Release/llvm-macosx-x86_64/./bin/clang++ -Wno-unknown-warning-option -Werror=unguarded-availability-new -fno-stack-protector -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -std=c++14 -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-class-memaccess -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wstring-conversion -fdiagnostics-color -Werror=switch -Wdocumentation -Wimplicit-fallthrough -Wunreachable-code -Woverloaded-virtual -DOBJC_OLD_DISPATCH_PROTOTYPES=0 -O3 -DNDEBUG -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -dynamiclib -Wl,-headerpad_max_install_names -dynamiclib\ -Wl,-headerpad_max_install_names -Xlinker -compatibility_version -Xlinker 1 -target x86_64-apple-macosx10.9 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -arch x86_64 -F /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/../../../Developer/Library/Frameworks -mmacosx-version-min=10.9 -Wl,-dead_strip  "-L/Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-perf/Ninja-Release/swift-macosx-x86_64/./lib/swift/macosx/x86_64" "-L/Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-perf/Ninja-Release/swift-macosx-x86_64/./bin/../lib/swift/macosx/x86_64" "-L/Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-perf/Ninja-Release/swift-macosx-x86_64/./bin/../lib/swift/macosx" -o lib/swift/macosx/x86_64/libswiftRemoteMirror.dylib -install_name @rpath/libswiftRemoteMirror.dylib stdlib/public/SwiftRemoteMirror/CMakeFiles/swiftRemoteMirror-macosx-x86_64.dir/SwiftRemoteMirror.cpp.o -L/Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-perf/Ninja-Release/llvm-macosx-x86_64/./lib lib/swift/macosx/x86_64/libswiftReflection.a && :
18:29:59 Undefined symbols for architecture x86_64:
18:29:59   "swift::Demangle::Node::dump()", referenced from:
18:29:59       (anonymous namespace)::Remangler::mangle(swift::Demangle::Node*) in libswiftReflection.a(Remangler.cpp.o)
18:29:59 ld: symbol(s) not found for architecture x86_64
```

nate-chandler commented:
@swift-ci please clean smoke benchmark

nate-chandler commented:
@swift-ci please clean smoke test

nate-chandler commented:
@swift-ci please clean smoke test

nate-chandler committed:
[metadata prespecialization] Create enum records.

Extracted implementation of SpecializedGenericStructMetadataBuilder into
SpecializedGenericNominalMetadataBuilderBase, a CRTP with a template
template argument for the CRTP superclass and a template argument for
the implementation.  That new type is now subclassed by
SpecializedGenericStructMetadataBuilder.  Additionally, this new type is
also subclassed by the newly added SpecializedGenericEnumMetadataBuilder
which is responsible for build the prespecialization of generic enum
metadata.

rdar://problem/56960887

nate-chandler commented:
@swift-ci please smoke test

nate-chandler commented:
@swift-ci please smoke test

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please smoke test

nate-chandler committed:
[metadata prespecialization] Direct refs to enums.

When a specialized usage of a generic enum occurs in the same module
where the enum was defined, directly reference the prespecialized
metadata.

rdar://problem/56994321

nate-chandler commented:
@swift-ci please smoke test

nate-chandler commented:
@swift-ci please smoke test

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci Please Test Source Compatibility

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test compiler performance

nate-chandler commented:
Release:

```
LLVM.NumLLVMBytesOutput | 1,923,558,944 | 1,920,178,772 | -3,380,172 | -0.18%
```

Debug:

```
LLVM.NumLLVMBytesOutput | 1,800,361,224 | 1,801,765,732 | 1,404,508 | 0.08%
```

Source:

https://ci.swift.org/job/swift-PR-compiler-performance-macOS/571/artifact/

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test compiler performance

nate-chandler commented:
Release:
```
LLVM.NumLLVMBytesOutput | 1,923,705,960 | 1,921,602,736 | -2,103,224 | -0.11%
```

Debug:
```
LLVM.NumLLVMBytesOutput | 1,800,469,268 | 1,802,125,694 | 1,656,426 | 0.09%
```

Source:

https://ci.swift.org/job/swift-PR-compiler-performance-macOS/576/artifact/comment.md

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler committed:
[DO NOT MERGE] Always enable prespecialized generic metadata.

Enabling the feature always allows the feature to be exercised by the
source compatibility suite.

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please smoke benchmark

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test

nate-chandler committed:
[metadata prespecialization] Standardize checks to ref tables.

Previously, some ad hoc checks were done in order to determine whether
the metadata access for a generic type was trivial.  Now, standard
predicates are used, specifically IRGenModule's member functions
isDependentConformance and isResilientConformance.

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please clean test linux platform

futurejones commented:
Can one of the admins verify this patch?
